bundles/org.openhab.binding.iaqualink/pom.xml

```diff
@@ -1,17 +1,58 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
-
   <modelVersion>4.0.0</modelVersion>
-
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
     <version>4.3.0-SNAPSHOT</version>
   </parent>
-
   <artifactId>org.openhab.binding.iaqualink</artifactId>
-
   <name>openHAB Add-ons :: Bundles :: iAquaLink Binding</name>
-
+  <properties>
+    <bnd.importpackage>!org.graalvm.nativeimage.hosted;!org.apache.tapestry5.json.*,!org.codehaus.jettison.json.*,!org.json.*,!com.fasterxml.jackson.*,!jakarta.json.*</bnd.importpackage>
+  </properties>
+  <dependencies>
+    <dependency>
+      <groupId>software.amazon.awssdk.iotdevicesdk</groupId>
```

Comment:
Maybe you can join @clinique to get rid of this SDK. For details see: https://github.com/openhab/openhab-addons/pull/16893#discussion_r1750860037

Reply:
I did actually have a look at using an alternative MQTT client, because it was originally such as pita to get the CRT working. The challenges that I had were:

 - Should support MQTTv5
 - Should work over Websockets
 - Should support AWS Signature V4 authentication

I'm sure that it's possible, but I doubt that it's simple. I would also question the longevity and applicability of what we were to build, given that AWS appear happy to build their own custom clients and protocol. We'd need to maintain a parallel pure-Java implementation.

I am sympathetic to the arguments regarding packaging the CRT, however given that AWS have such a strong position in IOT I would expect many bindings to have to interface with AWS, and hence wonder if a better solution would be to create a CRT OSGi for all bindings to use (instead of package themselves)?

Reply:
The alternative client to use would be the Paho MQTT client. It does look like it supports MQTTv5 over websockets, although not everything that AWS requires, for example: https://github.com/eclipse-paho/paho.mqtt.java/issues/830, https://github.com/eclipse-paho/paho.mqtt.java/issues/858 etc.

It should be coming, but it's not there yet.

Reply:
I searched a long time but found no alternative to get logged in with HiveMQ. There are very rare informations available when you try to escape AWS SDK.

Reply:
PAHO client seems abandoned. So there does not seem to be an alternative. I'm no expert in OSGI packages and if it would be usefull to add this SDK as OSGI package. Maybe @wborn can give some guidance?

Reply:
I've spend quite a while trying which Hive and Paho, and ironically I think I can get Paho to work, but no luck with Hive so far. Hive does not allow complete control over the query string escaping. @clinique what is the authentication mechanism that you need? This binding needs sigV4 over wss.

Reply:
Thje salus binding also has some experience with the AWS signing, not sure if it meets the demands of this PR.
PR is here: https://github.com/openhab/openhab-addons/pull/16840

```
<dependency>
      <groupId>software.amazon.awssdk.crt</groupId>
      <artifactId>aws-crt</artifactId>
      <version>0.29.19</version>
      <groupId>uk.co.lucasweb</groupId>
      <artifactId>aws-v4-signer-java</artifactId>
      <version>1.3</version>
      <scope>compile</scope>
</dependency>
```

Reply:
I did take a look at the library, but it wasn't sufficient for this binding as it doesn't support aws security tokens. However I'm already past this stage as I wrote my own signing code. I have managed to connect with Paho, however the problem with Hive is that it's not possible to provide a querystring containing a %2F (forward slash), as this is always decoded to / and not re-encoded. I will try to modify the signing code to sign based on decoded forward slashes, however I'm not sure if aws will allow this as it canonicalises the URLs and it may require a %2F.

Reply:
Nice, let's see how far you can get. Let me know if i can assist you in any way.

Reply:
@jpg0 , authentication is done [here](https://github.com/openhab/openhab-addons/pull/16893/files#diff-b8d3dd57e8803ec39109644d9d0bad57fe104997bcfe70b5a1476e08073e6535) - line 70. OAuth on worx is used (username/password) to get a token. The token is passed to AWS via MQTT on websocket (if I understood well).

Reply:
I can get all the tokens, the problem I have is constructing the signed URL to connect to the broker (over wss). This is for a sigV4 URL, you may be using another approach like mtls etc?

Reply:
Adding support for forward slashes to the hivemq client would probably best, but i can't asses how much work that or even if that is possible at all. Maybe best to create an issue for the specific hivemq here: https://github.com/hivemq/hivemq-mqtt-client

Reply:
> I can get all the tokens, the problem I have is constructing the signed URL to connect to the broker (over wss). This is for a sigV4 URL, you may be using another approach like mtls etc?

It's a customAuuthorizer : 
```
            MqttClientConnection connection = AwsIotMqttConnectionBuilder.newDefaultBuilder()
                    .withClientId("WX/USER/%s/%s/%s".formatted(userId, MQTT_USERNAME, productUuid))
        .withEndpoint(endpoint).withUsername(MQTT_USERNAME).withCleanSession(false).withKeepAliveSecs(300)
                    .withConnectionEventCallbacks(this).withWebsockets(true)
                    .withWebsocketHandshakeTransform(handshakeArgs -> {
                        HttpRequest httpRequest = handshakeArgs.getHttpRequest();
                        httpRequest.addHeader("x-amz-customauthorizer-name", AUTHORIZER_NAME);
                        httpRequest.addHeader("x-amz-customauthorizer-signature", tok[2]);
                        httpRequest.addHeader("jwt", tok[0] + "." + tok[1]);
                        handshakeArgs.complete(httpRequest);
                    }).build();
```

Reply:
@clinique That is certainly different to my case; I do not have all those fields (such as a username).

Reply:
@lsiepel I've spent a little more time on this and my conclusion is that we have a few choices:

- Ship the AWS SDK in the bundle. Possibly it can be shaved down a bit when shading, although given that much of it is JNI I'm not expecting much.
- Ship the AWS SDK as a shared bundle.
- Go with Hive: this means that we'd need to get a change into Hive, which unfortunately is down to an issue with the JDK URLEncoder, which doesn't escape forward slashes (or + signs). No chance of getting this changed there, and I'm afraid that Hive may be reluctant to deviate from it's behaviour. Seems some people say that escaping isn't required and others (e.g. Amazon) say that it is. The problem arises due to the signature calculation which is based on the URI. Whilst Hive allows you to pass a String, it puts in into a URI object (then gets it back out), which means that you cannot provide a raw querystring.
- Go with Paho: Whilst I have it working, it comes with 2 caveats: 1. It relies on [a commit](https://github.com/eclipse-paho/paho.mqtt.java/commit/fbf9544977c34ff46182e70bffac574021c6ac18) which whilst is about 4 years old, does not yet appear in an official release. 2. Paho seems somewhat abandoned, hence the first problem.

I also had a quick look to see if I could find any other suitable libraries. There is nothing in Java, the only thing close I found was https://github.com/davidepianca98/KMQTT/ which is Kotlin. Personally I'd love to see Kotlin shipped with openHAB, however that's another very heavy bundle.

Reply:
I would like @wborn's advice regarding the SDK. wdyt about adding it as a OSGi bundle?